swaybar/input.c

```diff
@@ -463,6 +464,115 @@ static const struct wl_touch_listener touch_listener = {
 	.orientation = wl_touch_orientation,
 };
 
+static uint32_t swipe_button_base(uint8_t fingers) {
+	if (fingers == 3) {
+		return SWAY_SWIPE_3;
+	} else if (fingers == 4) {
+		return SWAY_SWIPE_4;
+	}
+
+	return 0;
+}
+
+static void swipe_begin(void *data, struct zwp_pointer_gesture_swipe_v1 *wl_swipe,
+		uint32_t serial, uint32_t time, struct wl_surface *surface,
+		uint32_t fingers) {
+	struct swaybar_seat *seat = data;
+	struct swaybar_output *output;
+
+	wl_list_for_each(output, &seat->bar->outputs, link) {
+		if (output->surface == surface) {
+			surface = NULL;
+			break;
+		}
+	}
+
+	// we only want to respond to swipes on a bar surface
+	if (surface) {
+		return;
+	}
+
+	// Do not execute any button press bindings because due to the
+	// nature of swipes we cannot know yet what kind of swipe it is going
+	// to be.
+
+	struct swaybar_swipe *swipe = &seat->swipe;
+	swipe->swiping = true;
+	swipe->fingers = fingers;
+	swipe->travel_x = 0;
+	swipe->travel_y = 0;
+}
+
+static void swipe_update(void *data, struct zwp_pointer_gesture_swipe_v1 *wl_swipe,
+		uint32_t time, wl_fixed_t dx, wl_fixed_t dy) {
+	struct swaybar_seat *seat = data;
+	struct swaybar_swipe *swipe = &seat->swipe;
+
+	if (!swipe->swiping) {
+		sway_log(SWAY_DEBUG, "Ignoring swipe update without begin");
+		return;
+	}
+
+	swipe->travel_x += dx;
+	swipe->travel_y += dy;
+}
+
+static uint32_t swipe_to_button(struct swaybar_swipe *swipe) {
+	uint8_t direction;
+	if (fabs(swipe->travel_x) > fabs(swipe->travel_y)) {
+		if (swipe->travel_x > 0) {
+			direction = SWAY_SWIPE_DIR_RIGHT;
+		} else {
+			direction = SWAY_SWIPE_DIR_LEFT;
+		}
+	} else {
+		if (swipe->travel_y > 0) {
+			direction = SWAY_SWIPE_DIR_DOWN;
+		} else {
+			direction = SWAY_SWIPE_DIR_UP;
+		}
+	}
+
+	uint32_t button_base = swipe_button_base(swipe->fingers);
+	return button_base + direction;
+}
+
+static void swipe_end(void *data, struct zwp_pointer_gesture_swipe_v1
+		*zwp_pointer_gesture_swipe_v1, uint32_t serial, uint32_t time,
+		int32_t cancelled) {
+	struct swaybar_seat *seat = data;
+	struct swaybar_swipe *swipe = &seat->swipe;
+
+	if (!swipe->swiping) {
+		sway_log(SWAY_DEBUG, "Ignoring swipe end without begin");
+		return;
+	}
+
+	swipe->swiping = false;
```

Comment:
This isn't sufficient to track a swipe beginning to end.

Basically, it should not be possible to leave swaybar in an indefinite "swiping" state. If a user begins a swipe on the bar and moves the cursor away, the swiping state remains. Then if the user begins a new swipe and moves the cursor back to the bar before ending it, swaybar will execute a binding, but the begin and end events it saw actually belong to two different swipe events.

It was possible to trick clients this way before, but this PR also introduces another method by putting the swiping state in the default seatop (instead of it's own seatop), because now entering any non-default seatop will trash the swiping state in sway.

Either sway needs to be consistent and cancel the swipe in the compositor when we can no longer deliver the real end event, or clients should be consistent in understanding that not all swipes have an end.

The protocol isn't really clear which answer is correct and I don't know of any clients that actually use it anyway, so I guess we do whatever seems reasonable until told otherwise. I think the easiest solution here would be to cancel the ongoing swipe in swaybar in the pointer leave event. As a result, a swipe must begin and end on the bar surface to count.

Reply:
@RPigott I added the cancel to the pointer leave event as you suggested and can confirm that this avoids the "dangling" swipe state in swaybar. This is certainly cleaner than leaving the flag set indefinitely.

I was not able to reproduce the triggering of a binding by returning to the bar amidst a new swipe, though. Is it possible this is caused by the bar having its own layer?

Reply:
Hmm, I'm not sure why it wouldn't work for you but I'm pretty sure I didn't have trouble reproducing. It was a while ago and I didn't write down any other notes of my observations though. Make sure that sway itself isn't capturing the off-bar swipe since it will need to pass through the completion of that swipe to the bar at the end.

It shouldn't matter now since the new commit LGTM. Without the dangling swiping state I'm confident it's fixed. I'll try it out shortly.

Reply:
> Make sure that sway itself isn't capturing the off-bar swipe

Right, that was the reason. Once I disabled the swipe bindings in sway proper, I was able to reproduce the issue. Thanks!